### ISSUE TITLE
Emulator is calling https.callable functions remotely instead of from localhost

### ISSUE THREAD

**Issue by rmoskal** [firebase/firebase-tools#1528]
### [REQUIRED] Environment info

**firebase-tools:**

6.10.0

**Platform:**

Fedora 30

### [REQUIRED] Test case

I have a very simple firebase https.callable function:

```
const functions = require('firebase-functions');

const admin = require('firebase-admin');
admin.initializeApp();

exports.oauth = functions.https.onCall((data, ctx) => {

    return ({ok:1})

});
```

And in my calling code

```
const oauth = Firebase.functions().httpsCallable('oauth')
oauth('foo').then(console.log)
```

### [REQUIRED] Steps to reproduce

You just need to initialize FB all call the above in some front end code

### [REQUIRED] Expected behavior
After running firebase serve, I expected my front end code to call the function on localhost, as in the last line below:

```
✔  functions: Using node@8 from host.
✔  functions: Emulator started at http://localhost:5001
i  functions: Watching "/home/rob/development/stripmall/src/try1/web_app/functions" for Cloud Functions...
i  hosting: Serving hosting files from: build
✔  hosting: Local server: http://localhost:5000
i  Your code has been provided a "firebase-admin" instance.
**i  functions: HTTP trigger initialized at http://localhost:5001/stripmall-0001/us-central1/oauth**
```

### [REQUIRED] Actual behavior

Instead, my front end code is making CORS requests against the deployed functions url. But note, the url is malformed (undefined is  in the place where my project id should be). The result is the same whether I deploy the function or not.

> Access to fetch at 'https://us-central1-undefined.cloudfunctions.net/oauth' from origin 'http://localhost:5000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
> app.js:206 
> 

Firebase auth is working fine in the front end application.  The function is working fine in the shell.

**Comment by samtstern** [comment]
@rmoskal is this code in your web frontent?
```js
const oauth = Firebase.functions().httpsCallable('oauth')
oauth('foo').then(console.log)
```

It does not automatically know about the functions emulator.  Instead you have to tell it:
```js
Firebase.functions().useFunctionsEmulator('http://localhost:5001') 
```

**Comment by GriffinJohnston** [comment]
@samtstern Sorry to comment on a closed issue - it seems like a serious oversight that this doesn't appear anywhere in the functions emulator docs: https://firebase.google.com/docs/functions/local-emulator

**Comment by samtstern** [comment]
@GriffinJohnston thanks for the feedback!  Right now we document this behavior here:
https://firebase.google.com/docs/emulator-suite/connect_and_prototype

But you're right, it's confusing that the `/functions/local-emulator` docs are something else entirely.

**Comment by Elijer** [comment]
@samtstern Just to make sure I'm not making a dumb mistake, where does line get put:

Firebase.functions().useFunctionsEmulator('http://localhost:5001')

Somewhere in the public folder, like when the doc gets loaded? I'm putting it here:

```
document.addEventListener("DOMContentLoaded", event => {
  const app = firebase.app();
  const db = firebase.firestore();

  var functions = firebase.functions();
  // this following line is to make functions use the emulator
  //firebase.functions().useFunctionsEmulator("http://localhost:5001");
  var stripe = Stripe('pk_test_FjTxRNal2FWcwhlqw0WtIETQ00ZDxO3D9S');  
  document.getElementById("banner-login").innerText = "login";

  ```if (window.location.hostname === "localhost") {
    firebase.functions().useFunctionsEmulator("http://localhost:5001"); //wtfwtfwtfwtf
    console.log("localhost detected!");
    db.settings({
      host: "localhost:8080",
      ssl: false
    });
  }
});

```
And my https.callable functions are still getting called from the live server, no the emulators.

**Comment by GriffinJohnston** [comment]
@Elijer I put all this in an init file. So I'll do something like:
`const func = firebase.functions()

if (window.location.host.includes('localhost')) {
  func.useFunctionsEmulator('http://localhost:5001')
}

export { func, ... }
`

**Comment by Elijer** [comment]
@GriffinJohnston Thanks for the reply Griffin. In any case, I reset to my last clean commit and my code worked. What you're writing looks the same as what I wrote up above.

Putting the document.ready stuff in its own folder is maybe a good organizational technique though.

**Comment by samtstern** [comment]
@Elijer @GriffinJohnston that all looks correct!  If you're still having trouble with this please file a new issue.

**Comment by Alynva** [comment]
Oh my god, I spent so much time on this. I was getting just "Error: Response is not valid JSON object." because I don't deployed the function yet. I had to debug the source code of the lib to figure out that it was trying to use the cloud URL. The `useFunctionsEmulator` just works fine, but the [docs](https://firebase.google.com/docs/emulator-suite/connect_functions) of it are not linked anywhere in other languages.

**Comment by samtstern** [comment]
@Alynva sorry to hear that, are there other pages in the docs you think we should link to there from?

**Comment by Alynva** [comment]
@samtstern basically the submenus of "Connect your app and prototype" in the sidebar doesn't appear when other languages are selected. When you already are in one of these submenus and change the language, all the sidebar disappear. Actually, there are others pages where the sidebar just doesn't appear, for example [this](https://firebase.google.com/docs/firestore/manage-data/enable-offline), [this](https://firebase.google.com/docs/firestore/query-data/order-limit-data), and so on...

**Comment by samtstern** [comment]
@Alynva ah you're right, thank you for noticing!  I'll look into the translation pipeline.

**Comment by dragonsnapx** [comment]
Due to `useFunctionsEmulator` being deprecated, use `useEmulator("localhost", 5001)` instead.

**Comment by dmishh** [comment]
`firebase.functions().useEmulator('localhost', 5001)` is the solution, but if you have a different region than the default `us-central1`, you should call the httsCallable this way:

```
const httpsCallable = isRunningWithEmulator
  ? getFirebase() // for emulator
      .functions()
      .httpsCallable('yourHttpsCallable')
  : getFirebase() // for prod
      .app()
      .functions('<REGION>')
      .httpsCallable('yourHttpsCallable');

httpsCallable({ ... }).then((result) => ...);
```

**Comment by ottob** [comment]
Using the code:
```
getFirebase() // for emulator
      .functions()
      .httpsCallable('myFunction')
```
Did not work for me. It will then try to access my function with an url like: http://localhost:4002/myproject/us-central1/myFunction
The project, host and port is correct here since I called `useEmulator()` but the region is wrong. The function region should be `europe-west3` not `us-central1`.

And if I try to use the code below, it will not use localhost but will try to access the cloud at `https://europe-west3-myproject.cloudfunctions.net/myFunction` even if I have called useEmulator.
```
getFirebase() // for prod
      .app()
      .functions('europe-west3')
      .httpsCallable('myFunction')
```

The most natural thing would be if the `// for prod` code above could always be used and that it respected that useEmulator had been called.

My function is defined like this:
```
functions
  .region('europe-west3').https.onCall(async (data, context) => {
```

So I can't see a way to get this working with a custom region right now. ping @samtstern

**Comment by ad-si** [comment]
I'm having the same problem as @ottob. What's the solution here @samtstern?

**Comment by mheere** [comment]
Hi, I am also having the same issue as @ottob - does anyone know a way around this?

**Comment by rustinlewis** [comment]
Same problem.  Anyone find anything?

**Comment by Ben-Banso** [comment]
> Same problem. Anyone find anything?

For others and myself. You can enable the emulator on a specific region :

```
const firebaseApp = firebase.initializeApp(firebaseConfig);
if (location.hostname === 'localhost') {
  firebase.auth().useEmulator('http://localhost:8081/');
  firebase.firestore().useEmulator('localhost', 8080);
  firebase.functions().useEmulator('localhost', 5001);
  firebaseApp.functions("europe-west1").useEmulator('localhost', 5001);
}
```

**Comment by lasharela** [comment]
based on documentation `https://firebase.google.com/docs/emulator-suite/connect_functions` this needs to work:

import { getApp } from "firebase/app";
import { getFunctions, connectFunctionsEmulator } from "firebase/functions";

const functions = getFunctions(getApp());
connectFunctionsEmulator(functions, "localhost", 5001);

But it's not working.

**Comment by LogansUA** [comment]
> func.useFunctionsEmulator('http://localhost:5001')

Exactly. I'm having the same thing, it's just not working for me 😞
